### PR TITLE
Efficient cleanbots

### PR DESCRIPTION
**Pull request by Incoming5643**
I just saw a heard of cleanbots reduced to effectively one cleanbot because they all constantly targeted the same tiles. They thus ended up cleaning the same tile five times instead of multiple seperate tiles at once. This change will promote them "fanning out" properly.

### CODE REVIEW

**Comment by Cebasiz** [comment]
i don't support this, as it's a buff to sillicons and i think those are too strong already.

all of those poor janitors put out of a job by a swarm of AI-directed cleanbots. the AI is too strong already.

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 1]
```diff
@@ -80,9 +80,13 @@
...
 
 /mob/living/simple_animal/bot/cleanbot/process_scan(obj/effect/decal/cleanable/D)
+	. = list()
 	for(var/T in target_types)
 		if(istype(D, T))
-			return D
+			. += D
+	if(.)
+		return(pick(.))
+	return 0
```
Comment (duncathan) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]:
```
if(.)
    . = pick(.)
```
Replace the last three lines with that

Comment (Incoming5643) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: oh right, I forgot it was `.`

**Comment by Bawhoppen** [comment]
:+1:x:100:

**Comment by Incoming5643** [comment]
Another weird side effect of how things used to work: Cleanbots were always biased towards the south west thanks to how range works (the list is sorted  south to north and west to east)

**Comment by Dorsisdwarf** [comment]
Can you make this happen with other bots? It's annoying how after ten minutes, every bot on patrol is in a big herd.

**Comment by MrStonedOne** [comment]
Ya, bots generally tend to get grouped up at air locks.

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 2]
```diff
@@ -80,9 +80,12 @@
 			user << "<span class='danger'>[src] buzzes and beeps.</span>"
 
 /mob/living/simple_animal/bot/cleanbot/process_scan(obj/effect/decal/cleanable/D)
+	. = list()
 	for(var/T in target_types)
 		if(istype(D, T))
-			return D
+			. += D
+	if(.)
+		. = pick(.)
```
Comment (GunHog) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: Could you explain how this works? I have never seen this before.

Comment (duncathan) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: `.` is the default return value of every proc. It gets implicitly returned at the end if nothing else is.

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 3]
```diff
@@ -80,9 +80,12 @@
 			user << "<span class='danger'>[src] buzzes and beeps.</span>"
 
 /mob/living/simple_animal/bot/cleanbot/process_scan(obj/effect/decal/cleanable/D)
+	. = list()
 	for(var/T in target_types)
 		if(istype(D, T))
-			return D
+			. += D
+	if(.)
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: bad, ```.``` can be an empty list in this case, ```if(.)``` is always true because you set it to a list object above.
you also can't check the ```.len``` of it because it's not a list, a way around this would be ```if(.[1])``` to check that ```.``` atleast contains 1 element.

Comment (duncathan) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: `if(.:Len)`
:^)

Comment (duncathan) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: (Don't actually do this)

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: again, why bother with ```:len``` when we can abuse the range of the ```[]``` operator :P

Comment (duncathan) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: Won't it throw an OOB runtime if we do that?

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: I think byond handles that by returning false (out of bounds as false, that is) which is perfectly fine for us, but I will do a test.

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: Nope, it runtimes, and ```.:len``` doesn't even compile

So we're left with the current check, which doesn't even do what we need.
Shit.

Comment (duncathan) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]:
```
for(var/i in .)
    . = pick(.)
    break
```
How's that look, instead of the if check?

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: Loops and acting once have been proven to be faster than other approaches, so it certainly seems reasonable on the face of things.

but it does look hideous and would probably be best avoided by using a real list var for the entire proc, and then ```. = that_List.len ? pick(that_list) : null``` at the end

Comment (duncathan) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: Oh yea duh

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm (hunk 2 above)
Comment (GunHog) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: I know what that means, I am referring to the change he made. How does this cause cleanbots to avoid stepping over each other? It looks like you are just making a list and randomizing the tile selected? When I first made the comment, I thought you were trying to do some sort of shared list among the cleanbots.

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: The entire logic of this PR actually seems wrong.
If D is a subtype of T it is placed on the list, yet there is only one D, so if it was going to be picked, it was going to be picked before, except if it is somehow MORE than 1 of those types it'll be in the list more times which only results in a higher chance to pick D instead of... picking D.

tl;dr wtf @Incoming5643

**Comment by duncathan** [comment]
Literally everything here is wrong

**Comment by RemieRichards** [comment]
@duncathan 2 minutes two slow :P

**Comment by duncathan** [comment]
This always happens ;_;

**Comment by RemieRichards** [comment]
But glad to see I'm not the only one COMPLETELY BAMBOOZLED about how @Incoming5643's mind is working (or rather, isn't) in this PR.

**Comment by GunHog** [comment]
Whoa, go easy on the insults there, RR! We still need him for Datum Antags! We do not want him to get angry and ragequit before he gives us one of the greatest updates of 2016!

**Comment by Incoming5643** [comment]
I go to sleep and wake up to like 20 goddamn comments jesus fuck I just wanted the cleanbots to work better ;_:

**Comment by Incoming5643** [comment]
Also yeah I was apparently looking in the wrong place that proc only checks to make sure there's a cleanable decal in the tile. I ACTUALLY need to add this behavior in scan(), because that's where it's constantly picking the southwest corner. I will do that a little later.

:warning: This pull never happened :warning:
